subsys/console/tty.c

```diff
@@ -149,12 +160,54 @@ static int tty_getchar(struct tty_serial *tty)
 	return c;
 }
 
+static ssize_t tty_read_unbuf(struct tty_serial *tty, void *buf, size_t size)
+{
+	u8_t *p = buf;
+	size_t out_size = 0;
+	int res = 0;
+	u32_t timeout = tty->rx_timeout;
+
+	while (size) {
+		u8_t c;
+		res = uart_poll_in(tty->uart_dev, &c);
+		if (res <= -2) {
+			/* Error occured, best we can do is to return
+			 * accumulated data w/o error, or return error
+			 * directly if none.
+			 */
+			if (out_size == 0) {
+				errno = res;
```

Comment:
Two questions here:
1. where `errno` is defined?
2. Do you assume that tty module will not be multiinstance?

Reply:
@jarz-nordic, this is hard(ish) one.

1. First of all, it follows the pattern used by other code in this module.
2. And the code in this module does that, because the requirement is to be POSIX compatible.
3. POSIX defines `errno`, in `<errno.h>` (we define it "somewhere").
4. `errno` is a thread-local variable.

So, it wasn't that hard after all. The only realistic way to perform *concurrent* operations (and that's your concern) on multiple instances is via different threads, and POSIX has that covered. Otherwise, this API is not intended to be called from ISRs (but if one does, it might even not crash, just some aspects would be funky, like competing for errno writes).

Reply:
@pfalcon : Thank you for explanation.